doc/website/advanced/best-practice-for-testing.md

```diff
@@ -0,0 +1,263 @@
+
+Sometimes it is necessary to access a pointer returned by a method.
+Using an `EXPECT_*` to check for a `nullptr` is not sufficient if the pointer is dereferenced afterwards.
+It might crash the test application if it indeed is a `nullptr`.
+The implementation of the test object might be broken and only after the successful test run it can be assumed to be correct.
+To prevent a crash, the `nullptr` check has to be done with an `ASSERT_*`. This gracefully aborts the current test case and continues with the next one.
+The same applies to other potential dangerous operations, like accessing the value of an `cxx::optional` or `cxx::expected`.
+This check can be omitted if another test case already ensured that the test object operates as expected.
+
+Last but not least, apply the **DRY** principle (don't repeat yourself) and use typed and parameterized test to check multiple implementations and variations without repeating much code.
+
+# Practical Example
+
+Let's test the following class
+
+```c++
+class SingleDigitNumber
+{
+  public:
+    SingleDigitNumber() noexcept = default;
+
+    constexpr SingleDigitNumber(uint64_t number) noexcept
+        : m_number(number)
+    {
+    }
+
+    constexpr operator uint64_t() const noexcept
+    {
+        return m_number;
+    }
+
+    constexpr SingleDigitNumber operator+(const SingleDigitNumber rhs) const noexcept
+    {
+        return m_number + rhs.m_number;
+    }
+
+    constexpr SingleDigitNumber operator-(const SingleDigitNumber rhs) const noexcept
+    {
+        return m_number - rhs.m_number;
+    }
+
+  private:
+    uint64_t m_number;
+};
+```
+
+This test fixture will be used
+
+```c++
+class SingleDigitNumber_test : public Test
+{
+  public:
+    void SetUp() override{};
+    void TearDown() override{};
+};
+```
+
+## First Attempt
+
+Well, this is quite a simple class, so the tests must also be simple, right?
+
+```c++
+TEST_F(SingleDigitNumber_test, TestClass)
+{
+    SingleDigitNumber number1;
+    SingleDigitNumber number2(3U);
+    SingleDigitNumber number3(2U);
+
+    auto number4 = number2 + number3;
+    auto number5 = number2 - number3;
+
+    EXPECT_TRUE(number4 == 5U);
+    EXPECT_EQ(1U, static_cast<uint64_t>(number5));
+}
+```
+
+Done and we reached 100% coverage. We can clap ourselves on the shoulders and move on. Well, except that we can't.
+
+The test above has several major and minor flaws. The first thing that leaps to the eye, it doesn't follow the **AAA** pattern. When the test fails, we don't know why and have to look in the code to figure out what went wrong. But even if only one aspect of the class would have been tested, there would still be essentially just a `[  FAILED  ] SingleDigitNumber_test.TestClass` output and we would need to look at the code to know what exactly is broken. This test case checks too many aspects of the code and doesn't have a sensible name.
+
```

Comment:
The test is bad, no doubt, but we usually want to look at the name to get a rough idea why it failed and then we need to look which EXPECT failed. It is often infeasable to just have one EXPECT per test (e.g. an element is there so the container size is 1 and it is the element we put into the container and others). In other words: there must be some flexibility in what is "too many".

Reply:
I'll change `aspects` to `state transistions` and note that all side effects of that transition must be checked